proposals/esm-integration/README.md

```diff
@@ -1,3 +1,308 @@
+    count++;
+}
+
+export {count, incrementCount};
+```
+
+WebAssembly modules currently can't take part in these graphs. This means that developers who want to use WebAssembly modules have to manually trigger instantiation, which may not have completed by the time the JS module graph returns, breaking the mental model of modules.
+
+## Proposed solution
+
+Both goals can be addressed by enabling WebAssembly modules to be loaded using the ES module system. 
+
+This system has three phases:
+
+1. Construction — Module files are fetched and parsed and the module record is constructed.
+1. Instantiation — Exports and imports are wired up to memory locations. Functions are also initialized here.
+1. Evaluation — Top-level module code is evaluated so that non-function exports can be assigned their values.
+
+The work completed in these phases is defined in two different specs:
+
+- A host-specific specification (i.e. the HTML spec) defines how to load modules.
+- The ES module spec provides methods for loaders to use to parse, instantiate, and evaluate modules. The spec defines both an Abstract Module Record and a concrete subclass of this for JS modules, called the Source Text Module Record.
+
+This proposal will create a new WebAssembly Module subclass of the Abstract Module Record. For this, the 3 phases will be specified as follows (with the possible introduction of a sub-phase in Instantiation):
+
+### Construction
+
+During construction, all files for modules in the graph are downloaded and are parsed into Module Records. 
+
+Three things happen for each module during the Construction phase.
+
+1. Finding the file — Based on the module specifier, the host figures out where to get the file from using its module resolution algorithm. Module resolution algorithms differ between hosts. There is ongoing work to fix incompatibilities between Node.js and browser module resolution with the [package-name-maps](https://github.com/domenic/package-name-maps) proposal.
+1. Loading the file — The host fetches the file using the module path generated in module resolution. For browsers, the loading algorithm is specified in the [HTML spec](https://html.spec.whatwg.org/#fetch-a-module-script-tree). The WebAssembly MIME type will need to be added to this spec, but loading should largely work without change for WebAssembly modules. 
+1. Parsing the file — The runtime parses the file into a Module Record. For JavaScript, this work is specified in the [`ParseModule` method](https://tc39.github.io/ecma262/#sec-parsemodule) of the Source Text Module Record. This proposal will add a `ParseModule` method to the WebAssembly spec. The runtime creates the logical equivalent of `WebAssembly.Module` during this phase, and fills in the `RequestedModules` field of the module.
+
+### Instantiation
+
+During instantiation, exports and imports are wired up to each other. In JS, functions are initialized at this point. All other values are undefined until evaluation.
```

Comment:
I assume that here missing or mismatched imports cause the below initialization to fail?

Reply:
Good point, I had assumed that as well but should probably add a line to make it explicit.